pkg/server/cache/entrycache/fullcache.go

```diff
@@ -208,6 +254,43 @@ func (c *FullEntryCache) crawl(parentID spiffeID, seen map[spiffeID]struct{}) []
 	return entries
 }
 
+func (c *FullEntryCache) createOrUpdateEntry(commonEntry *common.RegistrationEntry) error {
+	protoEntry, err := api.RegistrationEntryToProto(commonEntry)
+	if err != nil {
+		return err
+	}
+
+	parentID := spiffeIDFromProto(protoEntry.ParentId)
+	cacheEntries := c.entries[parentID]
+
+	var i int
+	for i = 0; i < len(cacheEntries); i++ {
+		if cacheEntries[i].Id == protoEntry.Id {
+			cacheEntries[i] = protoEntry
+			break
+		}
+	}
+	if i == len(cacheEntries) {
+		cacheEntries = append(cacheEntries, protoEntry)
+	}
+
+	c.entries[parentID] = cacheEntries
+
+	return nil
+}
+
+func (c *FullEntryCache) deleteEntry(entryID string) {
+	for parentID, cacheEntries := range c.entries {
```

Comment:
The full cache wasn't designed to receive piece-meal updates so there are not any existing indices that make this efficient. I think we'll want to add a reverse lookup from entry ID to parentIDs as part of this change.